### PR TITLE
update client to support object instead of args

### PR DESCRIPTION
**Pull request by njmulsqb**
fixes zaproxy/zaproxy#7608

Signed-off-by: Najam Ul Saqib <najamulsaqib@tutamail.com>

### CODE REVIEW

**Comment by njmulsqb** [comment]
@thc202 Changelog has been updated, let me know if any other changes are required.

**Comment by njmulsqb** [comment]
Ping @kingthorin @psiinon

**Review by psiinon — APPROVED** [review]
I assume all of the relevant files were generated?
If so then it looks fine to me...

**Review by kingthorin — APPROVED** [review]

**Comment by thc202** [comment]
I'd have expected all files to be changed to support the new args but not seeing all of them (e.g. soap.js).
The `index.js` was not updated to provide the new APIs.

**Comment by kingthorin** [comment]
Thanks @thc202

File: src/index.js [hunk 1]
```diff
@@ -19,26 +19,36 @@
...
 const Autoupdate = require('./autoupdate');
 const Brk = require('./brk');
 const Context = require('./context');
 const Core = require('./core');
+const Exim = require('./exim');
 const ForcedUser = require('./forcedUser');
+const Graphql = require('./graphql');
 const HttpSessions = require('./httpSessions');
 const ImportLogFiles = require('./importLogFiles');
 const Importurls = require('./importurls');
```
Comment (thc202) [inline src/index.js:40]: These should be removed.

File: src/index.js [hunk 2]
```diff
@@ -19,26 +19,36 @@
...
+const Network = require('./network');
 const Openapi = require('./openapi');
 const Params = require('./params');
 const Pnh = require('./pnh');
 const Pscan = require('./pscan');
+const Reports = require('./reports');
 const Replacer = require('./replacer');
+const Retest = require('./retest');
+const Revisit = require('./revisit');
```
Comment (thc202) [inline src/index.js:49]: This should be after `Reveal`.

File: src/index.js [hunk 3]
```diff
@@ -64,25 +75,36 @@ function ClientApi(options) {
...
+  this.graphql = new Graphql(this);
   this.httpSessions = new HttpSessions(this);
   this.importLogFiles = new ImportLogFiles(this);
   this.importurls = new Importurls(this);
+  this.network = new Network(this);
   this.openapi = new Openapi(this);
   this.params = new Params(this);
   this.pnh = new Pnh(this);
   this.pscan = new Pscan(this);
+  this.reports = new Reports(this);
```
Comment (thc202) [inline src/index.js:102]: This should be after `replacer`.

**Comment by thc202** [comment]
The `spider` was not generated, opened a PR to fix that (zaproxy/zap-extensions#4408). It would be great if you used that branch to generate the code again and verify that the `spider` is generated with the latest core code.

**Comment by njmulsqb** [comment]
> The `spider` was not generated, opened a PR to fix that ([zaproxy/zap-extensions#4408](https://github.com/zaproxy/zap-extensions/pull/4408)). It would be great if you used that branch to generate the code again and verify that the `spider` is generated with the latest core code.

I generated the files after taking fresh pull including the changes you made, but it did not update the spider file.
https://paste.debian.net/1269918/

**Comment by thc202** [comment]
You need to apply this change again https://github.com/zaproxy/zap-api-nodejs/pull/26#issuecomment-1412046643

File: src/index.js (hunk 1 above)
Comment (thc202) [reply src/index.js]: To be clear the corresponding variables and files should also be removed.

**Review by njmulsqb — COMMENTED** [review]
Changes done and now spider file is also updating as well.

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,7 +4,32 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ## [Unreleased]
+### Added
+* Add the API of the following Add-Ons
```
Comment (thc202) [inline CHANGELOG.md:8]:
```suggestion
* Add the API of the following add-ons:
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,7 +4,32 @@ All notable changes to this project will be documented in this file.
...
+    * Network version 0.6.0
+    * Report Generation version 0.18.0
+    * Retest version 0.5.0
+    * Revisit version 4
+    * Wappalyzer - Technology Detection version 21.18.0
 
+### Changed
+* Update the API to support object as function parameters instead of individual parameters. [Github Issue](https://github.com/zaproxy/zaproxy/issues/7608)
+* Update Core APIs for 2.12.0
+* Update the APIs of following Add-Ons
```
Comment (thc202) [inline CHANGELOG.md:23]:
```suggestion
* Update the APIs of following add-ons:
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,7 +4,32 @@ All notable changes to this project will be documented in this file.
...
+    * Ajax Spider version 23.10.0
+    * OpenAPI version 31
+    * Plug-n-Hack Configuration version 13
+    * Replacer version 12
+    * Reveal version 5
+    * Selenium version 15.11.0
+    * SOAP Support version 16
+    * Spider version 0.2.0
+    * WebSockets version 28
```
Comment (thc202) [inline CHANGELOG.md:33]:
```suggestion

### Removed
* The APIs for the add-ons Import files containing URLs and Log File Importer were removed, superseded by Import/Export add-on.
```

File: src/users.js [hunk 7]
```diff
@@ -27,10 +27,14 @@ function Users(clientApi) {
   this.api = clientApi;
 }
 
-Users.prototype.usersList = function (contextid, callback) {
+/**
+ * Gets a list of users that belong to the context with the given ID, or all users if none provided.
+ * @param {string} contextid - The Context Id
```
Comment (thc202) [inline src/users.js:32]: Core could be regenerated again (or find and replace), the `Id` was changed to `ID` in all places.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**